core/src/main/java/com/linecorp/armeria/internal/DefaultTimeoutController.java

```diff
@@ -76,100 +84,166 @@ public void setTimeoutTask(TimeoutTask timeoutTask) {
+        lastExecutionTimeNanos = currentNanoTime;
+
+        if (newTimeoutMillis <= 0) {
+            invokeTimeoutTask();
+            state = State.TIMED_OUT;
+            return false;
         }
+
+        timeoutFuture = eventLoop.schedule(this::invokeTimeoutTask, newTimeoutMillis, TimeUnit.MILLISECONDS);
+        state = State.SCHEDULED;
+        return true;
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Note that the {@link TimeoutTask} should be set via the {@link #setTimeoutTask(TimeoutTask)} or
+     * the {@link #DefaultTimeoutController(TimeoutTask, EventLoop)} before calling this method.
+     *
+     * @return {@code true} if the current timeout is reset by the specified {@code newTimeoutMillis}.
+     *         {@code false} if the timeout has been triggered already or
+     *         the {@link TimeoutTask} could not be scheduled now.
+     */
     @Override
-    public void resetTimeout(long newTimeoutMillis) {
+    public boolean resetTimeout(long newTimeoutMillis) {
         ensureInitialized();
+        if (state == State.TIMED_OUT || !timeoutTask.canSchedule()) {
+            return false;
+        }
         if (newTimeoutMillis <= 0) {
             timeoutMillis = newTimeoutMillis;
             cancelTimeout();
-            return;
+            return true;
         }
 
-        if (timeoutTask.canSchedule()) {
-            final long currentNanoTime = System.nanoTime();
-            final long passedTimeMillis = TimeUnit.NANOSECONDS.toMillis(currentNanoTime - lastStartTimeNanos);
-            final long remainingTimeoutMillis = LongMath.saturatedSubtract(timeoutMillis, passedTimeMillis);
-            lastStartTimeNanos = currentNanoTime;
-            if (remainingTimeoutMillis == newTimeoutMillis) {
-                return;
-            }
+        // Cancel the previously scheduled timeout, if exists.
+        cancelTimeout();
+        timeoutMillis = newTimeoutMillis;
+        timeoutFuture = eventLoop.schedule(this::invokeTimeoutTask, newTimeoutMillis, TimeUnit.MILLISECONDS);
+        state = State.SCHEDULED;
+        return true;
+    }
 
-            // Cancel the previously scheduled timeout, if exists.
-            cancelTimeout();
-            timeoutMillis = newTimeoutMillis;
-            timeoutFuture = eventLoop.schedule(timeoutTask, newTimeoutMillis,
-                                               TimeUnit.MILLISECONDS);
+    /**
+     * Trigger the current timeout immediately.
+     *
+     * @return {@code true} if the current timeout is triggered successfully.
+     *         {@code false} if no timeout was scheduled previously, the timeout has been triggered already
+     *         or the {@link TimeoutTask} could not be scheduled now.
+     */
+    @Override
+    public boolean timeoutNow() {
+        checkState(timeoutTask != null,
+                   "setTimeoutTask(timeoutTask) is not called yet.");
+
+        if (State.SCHEDULED != state || !timeoutTask.canSchedule()) {
```

Comment:
Global comment - `state != ... or state == ...` for readability

Reply:
Sure! 😀